<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Introduction to R</title>
  <meta name="description" content="This is the complementary manual for QFAB’s Introduction to R workshop">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="Introduction to R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is the complementary manual for QFAB’s Introduction to R workshop" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Introduction to R" />
  
  <meta name="twitter:description" content="This is the complementary manual for QFAB’s Introduction to R workshop" />
  

<meta name="author" content="QFAB">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="more-information-on-r-and-bioconductor-for-genomics.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #0000ff; } /* Keyword */
code > span.ch { color: #008080; } /* Char */
code > span.st { color: #008080; } /* String */
code > span.co { color: #008000; } /* Comment */
code > span.ot { color: #ff4000; } /* Other */
code > span.al { color: #ff0000; } /* Alert */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #008000; font-weight: bold; } /* Warning */
code > span.cn { } /* Constant */
code > span.sc { color: #008080; } /* SpecialChar */
code > span.vs { color: #008080; } /* VerbatimString */
code > span.ss { color: #008080; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #0000ff; } /* ControlFlow */
code > span.op { } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ff4000; } /* Preprocessor */
code > span.do { color: #008000; } /* Documentation */
code > span.an { color: #008000; } /* Annotation */
code > span.cv { color: #008000; } /* CommentVar */
code > span.at { } /* Attribute */
code > span.in { color: #008000; } /* Information */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Abstract</a></li>
<li class="chapter" data-level="1" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>1</b> Getting started</a><ul>
<li class="chapter" data-level="1.1" data-path="how-to-read-this-book.html"><a href="how-to-read-this-book.html"><i class="fa fa-check"></i><b>1.1</b> How to read this book</a></li>
<li class="chapter" data-level="1.2" data-path="rstudio-ide.html"><a href="rstudio-ide.html"><i class="fa fa-check"></i><b>1.2</b> RStudio IDE</a></li>
<li class="chapter" data-level="1.3" data-path="alternatives-to-rstudio.html"><a href="alternatives-to-rstudio.html"><i class="fa fa-check"></i><b>1.3</b> Alternatives to RStudio</a></li>
<li class="chapter" data-level="1.4" data-path="the-r-workspace.html"><a href="the-r-workspace.html"><i class="fa fa-check"></i><b>1.4</b> The R workspace</a></li>
<li class="chapter" data-level="1.5" data-path="literate-programming.html"><a href="literate-programming.html"><i class="fa fa-check"></i><b>1.5</b> Literate programming</a></li>
<li class="chapter" data-level="1.6" data-path="documentation.html"><a href="documentation.html"><i class="fa fa-check"></i><b>1.6</b> Documentation</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-variables-objects-in-r.html"><a href="data-variables-objects-in-r.html"><i class="fa fa-check"></i><b>2</b> Data variables (objects) in R</a><ul>
<li class="chapter" data-level="2.1" data-path="declaring-variables.html"><a href="declaring-variables.html"><i class="fa fa-check"></i><b>2.1</b> Declaring variables</a></li>
<li class="chapter" data-level="2.2" data-path="best-practise-and-variable-naming-conventions.html"><a href="best-practise-and-variable-naming-conventions.html"><i class="fa fa-check"></i><b>2.2</b> Best practise and variable naming conventions</a></li>
<li class="chapter" data-level="2.3" data-path="data-types.html"><a href="data-types.html"><i class="fa fa-check"></i><b>2.3</b> Data types</a><ul>
<li class="chapter" data-level="2.3.1" data-path="data-types.html"><a href="data-types.html#numeric-and-integer"><i class="fa fa-check"></i><b>2.3.1</b> Numeric and Integer</a></li>
<li class="chapter" data-level="2.3.2" data-path="data-types.html"><a href="data-types.html#character"><i class="fa fa-check"></i><b>2.3.2</b> Character</a></li>
<li class="chapter" data-level="2.3.3" data-path="data-types.html"><a href="data-types.html#logical"><i class="fa fa-check"></i><b>2.3.3</b> Logical</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="determining-data-type.html"><a href="determining-data-type.html"><i class="fa fa-check"></i><b>2.4</b> Determining data type</a><ul>
<li class="chapter" data-level="2.4.1" data-path="determining-data-type.html"><a href="determining-data-type.html#typecasting"><i class="fa fa-check"></i><b>2.4.1</b> Typecasting</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>2.5</b> Functions</a><ul>
<li class="chapter" data-level="2.5.1" data-path="functions.html"><a href="functions.html#nesting-functions"><i class="fa fa-check"></i><b>2.5.1</b> Nesting functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>2.6</b> Data structures</a><ul>
<li class="chapter" data-level="2.6.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>2.6.1</b> Vectors</a></li>
<li class="chapter" data-level="2.6.2" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>2.6.2</b> Lists</a></li>
<li class="chapter" data-level="2.6.3" data-path="data-structures.html"><a href="data-structures.html#changing-vectors-and-lists"><i class="fa fa-check"></i><b>2.6.3</b> Changing Vectors and Lists</a></li>
<li class="chapter" data-level="2.6.4" data-path="data-structures.html"><a href="data-structures.html#factors"><i class="fa fa-check"></i><b>2.6.4</b> Factors</a></li>
<li class="chapter" data-level="2.6.5" data-path="data-structures.html"><a href="data-structures.html#matrices"><i class="fa fa-check"></i><b>2.6.5</b> Matrices</a></li>
<li class="chapter" data-level="2.6.6" data-path="data-structures.html"><a href="data-structures.html#data-frame"><i class="fa fa-check"></i><b>2.6.6</b> Data frame</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="finding-describing-and-removing-objects.html"><a href="finding-describing-and-removing-objects.html"><i class="fa fa-check"></i><b>2.7</b> Finding, describing and removing objects</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="file-and-data-inputoutput-io.html"><a href="file-and-data-inputoutput-io.html"><i class="fa fa-check"></i><b>3</b> File and data input/output (IO)</a><ul>
<li class="chapter" data-level="3.1" data-path="saving-variable-data.html"><a href="saving-variable-data.html"><i class="fa fa-check"></i><b>3.1</b> Saving variable data</a></li>
<li class="chapter" data-level="3.2" data-path="loading-variable-data.html"><a href="loading-variable-data.html"><i class="fa fa-check"></i><b>3.2</b> Loading variable data</a></li>
<li class="chapter" data-level="3.3" data-path="current-working-directory.html"><a href="current-working-directory.html"><i class="fa fa-check"></i><b>3.3</b> Current working directory</a><ul>
<li class="chapter" data-level="3.3.1" data-path="current-working-directory.html"><a href="current-working-directory.html#file-paths"><i class="fa fa-check"></i><b>3.3.1</b> File paths</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="importing-from-text-files.html"><a href="importing-from-text-files.html"><i class="fa fa-check"></i><b>3.4</b> Importing from text files</a></li>
<li class="chapter" data-level="3.5" data-path="writing-to-text-file.html"><a href="writing-to-text-file.html"><i class="fa fa-check"></i><b>3.5</b> Writing to text file</a></li>
<li class="chapter" data-level="3.6" data-path="other-io-operations.html"><a href="other-io-operations.html"><i class="fa fa-check"></i><b>3.6</b> Other IO operations</a><ul>
<li class="chapter" data-level="3.6.1" data-path="other-io-operations.html"><a href="other-io-operations.html#interactive-input"><i class="fa fa-check"></i><b>3.6.1</b> Interactive input</a></li>
<li class="chapter" data-level="3.6.2" data-path="other-io-operations.html"><a href="other-io-operations.html#reading-from-a-web-connection"><i class="fa fa-check"></i><b>3.6.2</b> Reading from a web connection</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="simple-data-analysis.html"><a href="simple-data-analysis.html"><i class="fa fa-check"></i><b>4</b> Simple data analysis</a><ul>
<li class="chapter" data-level="4.1" data-path="simple-data-cleaning.html"><a href="simple-data-cleaning.html"><i class="fa fa-check"></i><b>4.1</b> Simple Data cleaning</a></li>
<li class="chapter" data-level="4.2" data-path="summary-statistics.html"><a href="summary-statistics.html"><i class="fa fa-check"></i><b>4.2</b> Summary statistics</a><ul>
<li class="chapter" data-level="4.2.1" data-path="summary-statistics.html"><a href="summary-statistics.html#two-dimensional-datasets"><i class="fa fa-check"></i><b>4.2.1</b> Two-dimensional datasets</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="vectorisation.html"><a href="vectorisation.html"><i class="fa fa-check"></i><b>4.3</b> Vectorisation</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="packages-and-functions.html"><a href="packages-and-functions.html"><i class="fa fa-check"></i><b>5</b> Packages and functions</a><ul>
<li class="chapter" data-level="5.1" data-path="extending-r-with-packages.html"><a href="extending-r-with-packages.html"><i class="fa fa-check"></i><b>5.1</b> Extending R with packages</a></li>
<li class="chapter" data-level="5.2" data-path="user-defined-functions.html"><a href="user-defined-functions.html"><i class="fa fa-check"></i><b>5.2</b> User defined functions</a></li>
<li class="chapter" data-level="5.3" data-path="tips-for-functions.html"><a href="tips-for-functions.html"><i class="fa fa-check"></i><b>5.3</b> Tips for functions</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="more-data-analysis.html"><a href="more-data-analysis.html"><i class="fa fa-check"></i><b>6</b> More data analysis</a><ul>
<li class="chapter" data-level="6.1" data-path="apply.html"><a href="apply.html"><i class="fa fa-check"></i><b>6.1</b> apply()</a></li>
<li class="chapter" data-level="6.2" data-path="for-and-while-loops.html"><a href="for-and-while-loops.html"><i class="fa fa-check"></i><b>6.2</b> For and while loops</a><ul>
<li class="chapter" data-level="6.2.1" data-path="for-and-while-loops.html"><a href="for-and-while-loops.html#for-loops"><i class="fa fa-check"></i><b>6.2.1</b> For loops</a></li>
<li class="chapter" data-level="6.2.2" data-path="for-and-while-loops.html"><a href="for-and-while-loops.html#while-loop"><i class="fa fa-check"></i><b>6.2.2</b> While loop</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="visualising-data.html"><a href="visualising-data.html"><i class="fa fa-check"></i><b>7</b> Visualising data</a><ul>
<li class="chapter" data-level="7.1" data-path="boxplots.html"><a href="boxplots.html"><i class="fa fa-check"></i><b>7.1</b> Boxplots</a></li>
<li class="chapter" data-level="7.2" data-path="histograms.html"><a href="histograms.html"><i class="fa fa-check"></i><b>7.2</b> Histograms</a></li>
<li class="chapter" data-level="7.3" data-path="density-plots.html"><a href="density-plots.html"><i class="fa fa-check"></i><b>7.3</b> Density plots</a></li>
<li class="chapter" data-level="7.4" data-path="scatterplots.html"><a href="scatterplots.html"><i class="fa fa-check"></i><b>7.4</b> Scatterplots</a><ul>
<li class="chapter" data-level="7.4.1" data-path="scatterplots.html"><a href="scatterplots.html#scatterplot-matrix"><i class="fa fa-check"></i><b>7.4.1</b> Scatterplot matrix</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="bar-charts.html"><a href="bar-charts.html"><i class="fa fa-check"></i><b>7.5</b> Bar charts</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html"><i class="fa fa-check"></i>Appendix 1</a><ul>
<li class="chapter" data-level="" data-path="summary-of-object-types.html"><a href="summary-of-object-types.html"><i class="fa fa-check"></i>Summary of Object Types</a></li>
<li class="chapter" data-level="" data-path="more-information-on-r-and-bioconductor-for-genomics.html"><a href="more-information-on-r-and-bioconductor-for-genomics.html"><i class="fa fa-check"></i>More information on R and Bioconductor for Genomics</a></li>
<li class="chapter" data-level="7.6" data-path="going-further-with-r.html"><a href="going-further-with-r.html"><i class="fa fa-check"></i><b>7.6</b> Going further with R</a><ul>
<li class="chapter" data-level="7.6.1" data-path="going-further-with-r.html"><a href="going-further-with-r.html#vignettes"><i class="fa fa-check"></i><b>7.6.1</b> Vignettes</a></li>
<li class="chapter" data-level="7.6.2" data-path="going-further-with-r.html"><a href="going-further-with-r.html#citations"><i class="fa fa-check"></i><b>7.6.2</b> Citations</a></li>
<li class="chapter" data-level="7.6.3" data-path="going-further-with-r.html"><a href="going-further-with-r.html#sessioninfo"><i class="fa fa-check"></i><b>7.6.3</b> SessionInfo</a></li>
<li class="chapter" data-level="7.6.4" data-path="going-further-with-r.html"><a href="going-further-with-r.html#other-ways-of-running-r"><i class="fa fa-check"></i><b>7.6.4</b> Other ways of running R</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="going-further-with-r" class="section level2">
<h2><span class="header-section-number">7.6</span> Going further with R</h2>
<p>This section is a quick glance at the other features in R that will help you in your future data analysis. We will quickly look at:</p>
<ul>
<li><strong>Documentation</strong> with R</li>
<li><strong>Vignettes</strong> that come with R packages including citing packages</li>
<li><strong>SessionInfo</strong> information to reproduce your data analysis, and</li>
<li><strong>Other ways to run R</strong></li>
</ul>
<div id="vignettes" class="section level3">
<h3><span class="header-section-number">7.6.1</span> Vignettes</h3>
<p>Just now, we explored some of the help functions built into R. As well as this builtin help, most R packages also come with their own, often extensive, documentation. This is normally in the form of a vignette, a document that provides a task-oriented description of package functionality. Vignettes contain executable examples and are intended to be used interactively. You can also download the vignette for a package that isn’t installed on your system by visiting the CRAN, Bioconductor project webpages, or via an internet search.</p>
<div class="rmdexercise">
<p>
<strong>Using vignettes (optional)</strong>
</p>
<ol style="list-style-type: decimal">
<li>
<p>
The <code>vignette()</code> command provides access to vignettes for installed packages.
</p>
</li>
<li>
Run <code>vignette()</code> at the console to list all the vignettes available.
<ul>
<li>
Notice that a package can, but does not always, have a vignette with the package name - for example the <code>annotate</code> package has an annotate vignette, but <code>BiocParallel</code> only has the <code>IntroductionToBiocParallel</code> vignette
</li>
</ul>
</li>
<li>
<p>
Type <code>vignette(“biomaRt”)</code> to open the vignette for the biomaRt package, which links to the Biomart gene annotation database.
</p>
</li>
<li>
Some vignettes have non-unique names - for example, several packages have an intro vignette.
<ol style="list-style-type: lower-alpha">
<li>
Try <code>vignette(“intro”)</code>
</li>
<li>
To get the limma package vignette, use the command <code>vignette(“intro”, package = “limma”)</code>
</li>
</ol>
</li>
</ol>
</div>
</div>
<div id="citations" class="section level3">
<h3><span class="header-section-number">7.6.2</span> Citations</h3>
<p>The R core development team and the very active community of package authors have invested a lot of time and effort in creating R as it is today. Please give credit where credit is due and cite R and R packages when you use them for data analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">citation</span>(<span class="st">&quot;VennDiagram&quot;</span>)
<span class="kw">citation</span>(<span class="st">&quot;limma&quot;</span>)</code></pre></div>
</div>
<div id="sessioninfo" class="section level3">
<h3><span class="header-section-number">7.6.3</span> SessionInfo</h3>
<p>Congratulations - you have now conquered the basics of R and you are ready to start breaking things! One of the most useful aspects of R is the rich ecosystem of supplementary packagess that can aid, facilitate and empower your research. The Bioconductor framework contains hundreds of packages of methods of relevance to biologists working with biological data (population genetics, SNPs, NGS reads, microarrays, mass spectrometry etc). As we discovered in the Vignette section, there are well documented tutorials and guides that will work us through the application of fabulous methods and workflows. You will come across some unexpected ’error’ messages.</p>
<p>R packages are written by people like the QFAB bioinformaticians. Well intentioned, but busy. Code is crafted lovingly and tested in the scenarios that are implicit in our daily development. You are likely to be running R on a different computer with combinations of installed packages that are subtly different to those that we wrote the software with. If you identify a ”bug” in that the package gives a wrong result, fails with a hairy error message or misbehaves it is worthwhile to let the developer know that you are having a problem. Reproducibility of the error is critical - we would like to know which version of R is being used and all of the packages that are loaded in memory so that the developer can agree that yes, there is a problem and can help in the resolution.</p>
<p>It is trivial to report the data on the R environment that you are working in - the SessionInfo function reports installed packages and their versions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(limma)</code></pre></div>
<pre><code>## Loading required package: limma</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.4.1 (2017-06-30)
## Platform: x86_64-redhat-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so
## 
## locale:
##  [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
##  [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
##  [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] limma_3.30.13    kableExtra_0.4.0 knitr_1.17      
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.12     bookdown_0.5     codetools_0.2-15 digest_0.6.12   
##  [5] rprojroot_1.2    R6_2.2.2         backports_1.1.0  magrittr_1.5    
##  [9] evaluate_0.10.1  highr_0.6        httr_1.3.1       rlang_0.1.2     
## [13] stringi_1.1.5    rstudioapi_0.6   xml2_1.1.1       rmarkdown_1.6   
## [17] tools_3.4.1      stringr_1.2.0    readr_1.1.1      hms_0.3         
## [21] yaml_2.1.14      compiler_3.4.1   rvest_0.3.2      htmltools_0.3.6 
## [25] tibble_1.3.4</code></pre>
</div>
<div id="other-ways-of-running-r" class="section level3">
<h3><span class="header-section-number">7.6.4</span> Other ways of running R</h3>
<p>Today we are using RStudio, but R can be run in a number of different ways; for example, you could embed some R functionality in a computer program written in Python or Java. As discussed in sections <a href="rstudio-ide.html#rstudio-ide">1.2</a> and <a href="alternatives-to-rstudio.html#alternatives-to-rstudio">1.3</a>, both R and RStudio provide an interactive environment where you are prompted to enter a single command at the time. In this training course we are providing you with snippets of information that could be used within a data analysis workflow. These are intended to be run interactively.</p>
<p>In a bioinformatics laboratory the scientists who use R craft a mixture of packages and workflow scripts into analytical pipelines. R may be even be called by other software environments. Bioinformaticians will typically use the R console for refining, tuning and modifying existing scripts that they have written. The workflow is run through master scripts. In this section we will look at a simple R script that will create a table of information and write it out to file. We will be having a more complete look at the process of writing data to file in a later section (Section <a href="#reading-and-writing-tabular-data"><strong>??</strong></a>).</p>
<p>The advantage of running R scripts as a batch analysis is that larger and more complex analyses (such as the mapping of short DNA sequence reads) can be run overnight and each of the commands will run successively as prior commands complete.</p>
<div id="preparing-an-r-script" class="section level4">
<h4><span class="header-section-number">7.6.4.1</span> Preparing an R script</h4>
<p>An R script is a container for multiple R commands. It is intended to be run largely hands-off.</p>
<p>The RStudio software provides us with a very convenient way to create an R script - in the file dialog there is the option to create file and an R Script is a primary file type. R scripts typically have the extension <code>.R</code>. The file is a plain text file and needs to know the packages that should be loaded, the objects that should be set and the working environment where we should be working.</p>
</div>
<div id="running-an-r-script" class="section level4">
<h4><span class="header-section-number">7.6.4.2</span> Running an R script</h4>
<p>The following command is how you run an R script from a terminal window. This is outside of RStudio.</p>
<pre><code>R CMD BATCH [options] my_script.R [outfile]
R.exe&quot; CMD BATCH
--vanilla --slave &quot;c:\my projects\my_script.R&quot;</code></pre>
<p>We will demonstrate the next exercise to show you the method to run a script file using RStudio. You can repeat this exercise in your own time.</p>
<div class="rmdexercise">
<p>
<strong>Running a script in RStudio</strong>
</p>
<ul>
<li>
Create a new script file by going to <strong>File</strong> &gt; <strong>New File</strong> &gt; <strong>R Script</strong>.
</li>
<li>
Enter some commands in the editor window, for example the code chunk below.
</li>
<li>
Save the script with a file name e.g. “area-rectangle.R”
</li>
<li>
<p>
In RStudio, there are 3 ways you can run this script from top to bottom:
</p>
<ol style="list-style-type: decimal">
<li>
In the <strong>console</strong> window, type in the command: <code>source(“area-rectangle.R”)</code>. What is the output?
</li>
<li>
In the <strong>editor</strong> window, top right corner click on the <strong>arrow</strong> next to the <strong>Source</strong> button, select <strong>Source</strong>. What is the output?
</li>
<li>
Repeat step 2 but this time select <strong>Source with Echo</strong>. What is the difference between step 2 and 3?
</li>
</ol>
</li>
</ul>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">date</span>())
<span class="kw">print</span>(<span class="st">&quot;This is an R script!&quot;</span>)
length &lt;-<span class="st"> </span><span class="fl">10.25</span>
width &lt;-<span class="st"> </span><span class="fl">4.35</span>
area.of.rectangle &lt;-<span class="st"> </span>length *<span class="st"> </span>width
<span class="kw">print</span>(<span class="kw">paste</span>(<span class="st">&quot;The area of a rectangle with length=&quot;</span>,
              length, 
              <span class="st">&quot;cm by width=&quot;</span>,
              width,
              <span class="st">&quot;cm is&quot;</span>,
              area.of.rectangle,<span class="st">&quot;square cm.&quot;</span>))</code></pre></div>
<div class="rmdtip">
<p>
<strong>Use [Tab] to autocomplete</strong>
</p>
<p>
You can use the tab key to autocomplete a variable name, a function name or a filename (if the file is in the working directory). This will save you a lot of time and prevent typing errors when you are analysing your data.
</p>
</div>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="more-information-on-r-and-bioconductor-for-genomics.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": {}
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["intro-to-r.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

<!-- </html> -->
