<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Introduction to R</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="This is the complementary manual for QFAB’s Introduction to R workshop">
  <meta name="generator" content="bookdown 0.3 and GitBook 2.6.7">

  <meta property="og:title" content="Introduction to R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is the complementary manual for QFAB’s Introduction to R workshop" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Introduction to R" />
  
  <meta name="twitter:description" content="This is the complementary manual for QFAB’s Introduction to R workshop" />
  

<meta name="author" content="QFAB">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="current-working-directory.html">
<link rel="next" href="scan.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="css\style.css" type="text/css" />
</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Abstract</a></li>
<li class="chapter" data-level="1" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>1</b> Getting started</a><ul>
<li class="chapter" data-level="1.1" data-path="how-to-read-this-book.html"><a href="how-to-read-this-book.html"><i class="fa fa-check"></i><b>1.1</b> How to read this book</a></li>
<li class="chapter" data-level="1.2" data-path="rstudio-ide.html"><a href="rstudio-ide.html"><i class="fa fa-check"></i><b>1.2</b> RStudio IDE</a></li>
<li class="chapter" data-level="1.3" data-path="alternatives-to-rstudio.html"><a href="alternatives-to-rstudio.html"><i class="fa fa-check"></i><b>1.3</b> Alternatives to RStudio</a></li>
<li class="chapter" data-level="1.4" data-path="the-r-workspace.html"><a href="the-r-workspace.html"><i class="fa fa-check"></i><b>1.4</b> The R workspace</a></li>
<li class="chapter" data-level="1.5" data-path="literate-programming.html"><a href="literate-programming.html"><i class="fa fa-check"></i><b>1.5</b> Literate programming</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-variables-objects-in-r.html"><a href="data-variables-objects-in-r.html"><i class="fa fa-check"></i><b>2</b> Data variables (objects) in R</a><ul>
<li class="chapter" data-level="2.1" data-path="declaring-variables.html"><a href="declaring-variables.html"><i class="fa fa-check"></i><b>2.1</b> Declaring variables</a></li>
<li class="chapter" data-level="2.2" data-path="datatypes.html"><a href="datatypes.html"><i class="fa fa-check"></i><b>2.2</b> Datatypes</a><ul>
<li class="chapter" data-level="2.2.1" data-path="datatypes.html"><a href="datatypes.html#numeric"><i class="fa fa-check"></i><b>2.2.1</b> Numeric</a></li>
<li class="chapter" data-level="2.2.2" data-path="datatypes.html"><a href="datatypes.html#integer"><i class="fa fa-check"></i><b>2.2.2</b> Integer</a></li>
<li class="chapter" data-level="2.2.3" data-path="datatypes.html"><a href="datatypes.html#logical"><i class="fa fa-check"></i><b>2.2.3</b> Logical</a></li>
<li class="chapter" data-level="2.2.4" data-path="datatypes.html"><a href="datatypes.html#character"><i class="fa fa-check"></i><b>2.2.4</b> Character</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>2.3</b> Functions</a><ul>
<li class="chapter" data-level="2.3.1" data-path="functions.html"><a href="functions.html#nesting-functions"><i class="fa fa-check"></i><b>2.3.1</b> Nesting functions</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>2.4</b> Data structures</a><ul>
<li class="chapter" data-level="2.4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>2.4.1</b> Vectors</a></li>
<li class="chapter" data-level="2.4.2" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>2.4.2</b> Lists</a></li>
<li class="chapter" data-level="2.4.3" data-path="data-structures.html"><a href="data-structures.html#changing-vectors-and-lists"><i class="fa fa-check"></i><b>2.4.3</b> Changing Vectors and Lists</a></li>
<li class="chapter" data-level="2.4.4" data-path="data-structures.html"><a href="data-structures.html#factors"><i class="fa fa-check"></i><b>2.4.4</b> Factors</a></li>
<li class="chapter" data-level="2.4.5" data-path="data-structures.html"><a href="data-structures.html#matrices"><i class="fa fa-check"></i><b>2.4.5</b> Matrices</a></li>
<li class="chapter" data-level="2.4.6" data-path="data-structures.html"><a href="data-structures.html#data-frame"><i class="fa fa-check"></i><b>2.4.6</b> Data frame</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="finding-describing-and-removing-objects.html"><a href="finding-describing-and-removing-objects.html"><i class="fa fa-check"></i><b>2.5</b> Finding, describing and removing objects</a></li>
<li class="chapter" data-level="2.6" data-path="best-practise-and-variable-naming-conventions.html"><a href="best-practise-and-variable-naming-conventions.html"><i class="fa fa-check"></i><b>2.6</b> Best practise and variable naming conventions</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="going-further-with-r.html"><a href="going-further-with-r.html"><i class="fa fa-check"></i><b>3</b> Going further with R</a><ul>
<li class="chapter" data-level="3.1" data-path="documentation.html"><a href="documentation.html"><i class="fa fa-check"></i><b>3.1</b> Documentation</a></li>
<li class="chapter" data-level="3.2" data-path="extending-r-with-packages.html"><a href="extending-r-with-packages.html"><i class="fa fa-check"></i><b>3.2</b> Extending R with packages</a></li>
<li class="chapter" data-level="3.3" data-path="vignettes.html"><a href="vignettes.html"><i class="fa fa-check"></i><b>3.3</b> Vignettes</a><ul>
<li class="chapter" data-level="3.3.1" data-path="vignettes.html"><a href="vignettes.html#citations"><i class="fa fa-check"></i><b>3.3.1</b> Citations</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="sessioninfo.html"><a href="sessioninfo.html"><i class="fa fa-check"></i><b>3.4</b> SessionInfo</a></li>
<li class="chapter" data-level="3.5" data-path="other-ways-of-running-r.html"><a href="other-ways-of-running-r.html"><i class="fa fa-check"></i><b>3.5</b> Other ways of running R</a><ul>
<li class="chapter" data-level="3.5.1" data-path="other-ways-of-running-r.html"><a href="other-ways-of-running-r.html#preparing-an-r-script"><i class="fa fa-check"></i><b>3.5.1</b> Preparing an R script</a></li>
<li class="chapter" data-level="3.5.2" data-path="other-ways-of-running-r.html"><a href="other-ways-of-running-r.html#running-an-r-script"><i class="fa fa-check"></i><b>3.5.2</b> Running an R script</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="importing-and-exporting-data-from-r.html"><a href="importing-and-exporting-data-from-r.html"><i class="fa fa-check"></i><b>4</b> Importing and exporting data from R</a><ul>
<li class="chapter" data-level="4.1" data-path="saving-session-data.html"><a href="saving-session-data.html"><i class="fa fa-check"></i><b>4.1</b> Saving session data</a></li>
<li class="chapter" data-level="4.2" data-path="current-working-directory.html"><a href="current-working-directory.html"><i class="fa fa-check"></i><b>4.2</b> Current working directory</a><ul>
<li class="chapter" data-level="4.2.1" data-path="current-working-directory.html"><a href="current-working-directory.html#file-paths"><i class="fa fa-check"></i><b>4.2.1</b> File paths</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="reading-and-writing-tabular-data.html"><a href="reading-and-writing-tabular-data.html"><i class="fa fa-check"></i><b>4.3</b> Reading and writing tabular data</a></li>
<li class="chapter" data-level="4.4" data-path="scan.html"><a href="scan.html"><i class="fa fa-check"></i><b>4.4</b> Scan</a></li>
<li class="chapter" data-level="4.5" data-path="reading-from-a-web-connection.html"><a href="reading-from-a-web-connection.html"><i class="fa fa-check"></i><b>4.5</b> Reading from a web connection</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="transforming-and-summarising-data-in-r.html"><a href="transforming-and-summarising-data-in-r.html"><i class="fa fa-check"></i><b>5</b> Transforming and summarising data in R</a><ul>
<li class="chapter" data-level="5.1" data-path="user-defined-functions.html"><a href="user-defined-functions.html"><i class="fa fa-check"></i><b>5.1</b> User defined functions</a></li>
<li class="chapter" data-level="5.2" data-path="for-loops-while-loops-and-vectorisation.html"><a href="for-loops-while-loops-and-vectorisation.html"><i class="fa fa-check"></i><b>5.2</b> For loops, While loops and Vectorisation</a></li>
<li class="chapter" data-level="5.3" data-path="apply-lapply.html"><a href="apply-lapply.html"><i class="fa fa-check"></i><b>5.3</b> Apply, Lapply</a></li>
<li class="chapter" data-level="5.4" data-path="basic-data-analysis-methods.html"><a href="basic-data-analysis-methods.html"><i class="fa fa-check"></i><b>5.4</b> Basic data analysis methods</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-presentation-and-reporting.html"><a href="data-presentation-and-reporting.html"><i class="fa fa-check"></i><b>6</b> Data presentation and reporting</a><ul>
<li class="chapter" data-level="6.1" data-path="boxplots.html"><a href="boxplots.html"><i class="fa fa-check"></i><b>6.1</b> Boxplots</a><ul>
<li class="chapter" data-level="6.1.1" data-path="boxplots.html"><a href="boxplots.html#native"><i class="fa fa-check"></i><b>6.1.1</b> Native</a></li>
<li class="chapter" data-level="6.1.2" data-path="boxplots.html"><a href="boxplots.html#ggplot2geom_boxplot"><i class="fa fa-check"></i><b>6.1.2</b> ggplot2::geom_boxplot()</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="histograms.html"><a href="histograms.html"><i class="fa fa-check"></i><b>6.2</b> Histograms</a><ul>
<li class="chapter" data-level="6.2.1" data-path="histograms.html"><a href="histograms.html#native-1"><i class="fa fa-check"></i><b>6.2.1</b> Native</a></li>
<li class="chapter" data-level="6.2.2" data-path="histograms.html"><a href="histograms.html#ggplot2---geom_histogram"><i class="fa fa-check"></i><b>6.2.2</b> ggplot2 - geom_histogram()</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="bar-charts.html"><a href="bar-charts.html"><i class="fa fa-check"></i><b>6.3</b> Bar charts</a><ul>
<li class="chapter" data-level="6.3.1" data-path="bar-charts.html"><a href="bar-charts.html#native-2"><i class="fa fa-check"></i><b>6.3.1</b> Native</a></li>
<li class="chapter" data-level="6.3.2" data-path="bar-charts.html"><a href="bar-charts.html#ggplot2geom_bar"><i class="fa fa-check"></i><b>6.3.2</b> ggplot2::geom_bar()</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="scatterplots.html"><a href="scatterplots.html"><i class="fa fa-check"></i><b>6.4</b> Scatterplots</a><ul>
<li class="chapter" data-level="6.4.1" data-path="scatterplots.html"><a href="scatterplots.html#native-3"><i class="fa fa-check"></i><b>6.4.1</b> Native</a></li>
<li class="chapter" data-level="6.4.2" data-path="scatterplots.html"><a href="scatterplots.html#ggplot2geom_point"><i class="fa fa-check"></i><b>6.4.2</b> ggplot2::geom_point()</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html"><i class="fa fa-check"></i>Appendix 1</a><ul>
<li class="chapter" data-level="" data-path="summary-of-object-types.html"><a href="summary-of-object-types.html"><i class="fa fa-check"></i>Summary of Object Types</a></li>
<li class="chapter" data-level="" data-path="more-information-on-r-and-bioconductor-for-genomics.html"><a href="more-information-on-r-and-bioconductor-for-genomics.html"><i class="fa fa-check"></i>More information on R and Bioconductor for Genomics</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-2-solutions-to-exercises.html"><a href="appendix-2-solutions-to-exercises.html"><i class="fa fa-check"></i>Appendix 2 - Solutions to exercises</a><ul>
<li class="chapter" data-level="" data-path="using-vectors.html"><a href="using-vectors.html"><i class="fa fa-check"></i>2.4.1 Using vectors</a></li>
<li class="chapter" data-level="" data-path="using-lists.html"><a href="using-lists.html"><i class="fa fa-check"></i>2.4.3 Using lists</a></li>
<li class="chapter" data-level="" data-path="adding-factor-values.html"><a href="adding-factor-values.html"><i class="fa fa-check"></i>2.4.4 Adding factor values</a></li>
<li class="chapter" data-level="" data-path="creating-a-data-frame.html"><a href="creating-a-data-frame.html"><i class="fa fa-check"></i>2.4.6 Creating a data frame</a></li>
<li class="chapter" data-level="" data-path="reading-in-the-geneatlas-data.html"><a href="reading-in-the-geneatlas-data.html"><i class="fa fa-check"></i>4.2 Reading in the GeneAtlas data</a></li>
<li class="chapter" data-level="" data-path="creating-and-using-functions.html"><a href="creating-and-using-functions.html"><i class="fa fa-check"></i>5.1 Creating and using functions</a></li>
<li class="chapter" data-level="" data-path="loops-and-vectorisation.html"><a href="loops-and-vectorisation.html"><i class="fa fa-check"></i>5.2 Loops and vectorisation</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-3-the-sessioninfo-for-this-version-of.html"><a href="appendix-3-the-sessioninfo-for-this-version-of.html"><i class="fa fa-check"></i>Appendix 3 - The SessionInfo for this version of</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="reading-and-writing-tabular-data" class="section level2">
<h2><span class="header-section-number">4.3</span> Reading and writing tabular data</h2>
<p>Data is most commonly shared in tabular formats (Excel presents data in a tabular format). These are most-often text files that contain columns of data separated with comma, tab or other field-delimiters. In R there are some simple methods for quickly reading in massive amounts of data.</p>
<p>For the next few steps in this course we are going to explore a public dataset of human tissue gene expression data that was originally produced by Novartis. The dataset is known as the Novartis GeneAtlas data (now called BioGPS) and can be found in the GEO database under accession GSE1133 (<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE1133" class="uri">http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE1133</a>). We will look at some Affymetrix gene expression data from the study. These data have been prepared from microarray hybridisations using the Affymetrix Human Genome U133A Array. The data is in the file <code>data/Intro_to_R/GSE1133-GPL96_series_matrix.txt</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read data from the training server</span>
data.path &lt;-<span class="st"> &quot;data/Intro_to_R&quot;</span>

<span class="co"># Of course, you will need a different path if the data is on your local P.C.</span>
<span class="co"># e.g. data.path &lt;- &quot;C:/Intro_to_R/Data&quot;</span>
<span class="co"># Create a variable pointing to the file location.</span>

gene.atlas.file &lt;-<span class="st"> </span><span class="kw">file.path</span>(data.path, <span class="st">&quot;GSE1133-GPL96_series_matrix.txt&quot;</span>)

<span class="co"># Then look at the start and finish of that file</span>
<span class="kw">head</span>(<span class="kw">readLines</span>(gene.atlas.file))
<span class="kw">tail</span>(<span class="kw">readLines</span>(gene.atlas.file))</code></pre></div>
<p><strong>What is happening here?</strong></p>
<table>
<colgroup>
<col width="15%" />
<col width="84%" />
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>readLines()</code></td>
<td>reads each line from a target file into a vector. See also <code>read.table</code>, <code>read.csv</code> functions.</td>
</tr>
<tr class="even">
<td><code>file.path()</code></td>
<td><em>DOES NOT</em> read in the contents of a file. It just point to the location of a files on your system. You need to use <code>readLines</code>, <code>read.table</code> or similar to access the contents of that file.</td>
</tr>
</tbody>
</table>
<p>Using the <code>head</code> and <code>tail</code> commands helps us understand the nature of the file. It seems that the <code>!</code> prefix indicates metadata rather than data.</p>
<div class="rmdcaution">
<p>
Beware of printing large variables with RStudio Server, you can freeze your session. Judiciously use methods to limit the output:
</p>
<ul>
<li>
read into a variable first
</li>
<li>
limit the output
</li>
<li>
use <code>head()</code> / <code>tail()</code> functions
</li>
</ul>
</div>
<p>We have started to explore the contents of the expression data file using the code above, but at the moment have not imported it into a variable in R. Before we get to that stage, we will first load the annotation data associated with this expression dataset.</p>
<div class="rmdexercise">
<p>
<strong>Reading in tabular data</strong>
</p>
<ol style="list-style-type: decimal">
<li>
<p>
If you have not already done so, use the previous code chunk to set up your data.path and take an initial look at the expression data in the variable <code>gene.atlas.file</code>.
</p>
</li>
<li>
<p>
Follow the annotated code below to use <code>read.table()</code> to import the annotation data, or metadata, for this experiment
</p>
</li>
</ol>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># As before, define the location of the file</span>
annotation.file &lt;-<span class="st"> </span><span class="kw">file.path</span>(data.path,<span class="st">&quot;GSK_RNA.sdrf&quot;</span>)

<span class="co"># Look at the first six lines of the data</span>
<span class="co"># sep = &quot;\t&quot; means that the data is tab-separated</span>
<span class="kw">head</span>(<span class="kw">read.table</span>(annotation.file, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>))

<span class="co"># The last command included the header as the first line of data.</span>
<span class="co"># Specify (header = TRUE) that the first line contains header information</span>
<span class="kw">head</span>(<span class="kw">read.table</span>(annotation.file, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header=</span><span class="ot">TRUE</span>))

<span class="co"># Use read.table to import the data into an variable</span>
expressionMetadata &lt;-<span class="st"> </span><span class="kw">read.table</span>(annotation.file, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header=</span><span class="ot">TRUE</span>)

<span class="co"># Find out some information about the structure and contents of that new variable</span>
<span class="kw">tail</span>(expressionMetadata)
<span class="kw">dim</span>(expressionMetadata)
<span class="kw">colnames</span>(expressionMetadata)
<span class="kw">str</span>(expressionMetadata$Material.Type)

<span class="co"># By default, read.table imports text fields as factors. That&#39;s probably</span>
<span class="co"># not what we want, so disable it with stringsAsFactors=FALSE</span>
expressionMetadata &lt;-<span class="st"> </span><span class="kw">read.table</span>(annotation.file, 
                                 <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header=</span><span class="ot">TRUE</span>, 
                                 <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
<span class="kw">str</span>(expressionMetadata$Material.Type)</code></pre></div>
<p>Note that the lines starting with the default comment symbol (#) have been omitted from the import.</p>
<div class="rmdexercise">
<p>
<strong>Read the expression data into a variable called gene.atlas</strong>
</p>
<p>
In the last couple of sections we have used <code>readLines()</code> to look at the content of a file and we have used <code>read.table()</code> to read some experimental metadata into a data frame. Your challenge is now to import the “GSE1133-GPL96_series_matrix.txt” data into a data frame variable called <code>gene.atlas</code>.
</p>
<p>
<em>Hints</em>: * Use the <code>read.table</code> documentation to discover the nuances of the function. * This import is not as simple as the annotation file. You will need to make some inferences about the structure of the file (<em>Hint</em>: use <code>head()</code> and <code>tail()</code> and look at some of the options for <code>read.table()</code> in the Help.) * Make sure that the proper column names are included in the data frame.
</p>
</div>
<pre><code>##      ID_REF GSM18865 GSM18866 GSM18867 GSM18868 GSM18869 GSM18870 GSM18871
## 1 1007_s_at   2082.0   1577.2    231.8    328.4     83.6     73.0     53.2
## 2   1053_at    598.4    256.2     57.0    102.1     39.1     43.6     73.7
## 3    117_at    194.2    135.9    512.3    713.4    108.6     75.7    165.7
## 4    121_at   1910.5   1516.2    529.8    990.2    383.4    396.4    227.7
## 5 1255_g_at    138.4     53.8     30.7     75.7     16.9     22.8      4.1
## 6   1294_at    263.4    287.2    347.8    345.6    300.8    377.7    234.2
##   GSM18872 GSM18873 GSM18874 GSM18875 GSM18876 GSM18877 GSM18878 GSM18879
## 1     70.0     92.2     75.0     78.8     60.1    117.7    149.4    105.8
## 2     32.9     37.6     38.4     43.0     82.6     30.2      9.0     29.6
## 3    152.6     71.1    105.3    140.2     82.4     50.0     11.1     38.2
## 4    251.6    260.0    260.1    287.3    270.6    269.8    395.2    340.2
## 5      4.9     23.3     14.7     27.0      4.2     27.8      9.0     17.6
## 6    256.3    245.1    171.7    369.3    287.0    382.3    424.4    499.9
##   GSM18880 GSM18881 GSM18882 GSM18883 GSM18884 GSM18885 GSM18886 GSM18887
## 1     97.6    109.6    113.9     42.3     83.9     76.4     93.1    247.8
## 2     15.4     38.0     57.5     76.5     71.5     55.1     87.5    133.1
## 3     19.4     38.8      8.5     34.4      9.6     10.6     32.6     83.5
## 4    342.9    215.3    159.7    257.2    334.5    227.4    216.5    627.6
## 5      7.3      6.9      9.3      6.4      6.4      4.2     19.1     39.0
## 6    440.8    263.9    159.2    171.4    104.7    258.8    221.6    255.6
##   GSM18888 GSM18889 GSM18890 GSM18891 GSM18892 GSM18893 GSM18894 GSM18895
## 1    277.9    405.0    351.5    296.1    463.6    225.6    252.9    210.2
## 2    140.9    196.8    214.7     63.9     50.6     58.6     40.1     67.1
## 3     85.3     67.4     72.9    139.3     79.7     60.7     66.5     64.3
## 4    806.5    245.0    312.7   1038.8    951.5    632.7    620.8    797.9
## 5     60.5      6.8     19.3     53.0     40.2     34.0     35.7     41.5
## 6    287.4    188.6    183.3    201.4    193.9    135.0    103.7    106.3
##   GSM18896 GSM18897 GSM18898 GSM18899 GSM18900 GSM18901 GSM18902 GSM18903
## 1     98.8    302.0    358.2   1008.2    806.8    499.6    554.9    387.5
## 2    106.3    177.6    120.5     72.0     93.0     89.1    111.1     58.7
## 3     79.0     84.6     93.3    145.3    163.9    217.2    115.6    159.4
## 4    601.5    703.2    749.1    991.6   1042.0   1100.6   1022.6   1163.6
## 5     44.8     60.1     45.4     54.6     59.3     47.2     67.6     40.7
## 6    103.5    151.9    189.0    396.3    430.4    397.6    302.8    421.0
##   GSM18904 GSM18905 GSM18906 GSM18907 GSM18908 GSM18909 GSM18910 GSM18911
## 1    404.8   1195.7    713.4     91.3     30.9    518.4    598.4   1442.0
## 2     65.6    108.8     53.3    141.7     90.7    119.3     51.9     23.5
## 3    235.2    403.9    212.1     99.2     85.2    278.6    148.8    223.4
## 4   1314.8   2879.3   1847.8    722.2    541.6   1924.4   2362.2   2004.0
## 5     58.1    145.5     18.7     45.4      1.7     86.4     95.5    410.7
## 6    408.0    463.3    358.8    135.6    118.9    451.9    407.7    475.6
##   GSM18912 GSM18913 GSM18914 GSM18915 GSM18916 GSM18917 GSM18918 GSM18919
## 1   1400.8   2076.0   2644.9   1978.3   2182.1   2784.9   2738.3   2855.6
## 2     26.5     32.5     94.0    108.0     51.5     33.0     96.8     61.3
## 3     71.2    351.3    427.5    306.2    310.6    213.6    365.9    205.8
## 4   2451.0   2449.2   2026.1   1901.9   1950.4   3687.7   3131.7   2105.6
## 5    450.0    104.6    360.4    109.2    140.8    110.3    196.5    242.4
## 6    388.5    873.8    973.5    406.1    422.3    624.2    518.2    527.4
##   GSM18920 GSM18921 GSM18922 GSM18923 GSM18924 GSM18925 GSM18926 GSM18927
## 1   3351.4   2390.5   2837.6   3499.5   3183.4   2725.0   2825.7   2865.1
## 2     78.4     69.8     93.0     50.8     72.4     87.7     27.3    103.6
## 3    120.3    188.8    183.4    458.4    396.5    179.7    300.7    222.0
## 4   2231.3   1538.6   1975.7   2533.2   2679.1   2100.0   2223.5   2146.3
## 5    319.3     87.9    192.3    230.1    244.3    261.7    349.6    226.5
## 6    406.6    270.5    361.1    581.4    637.2    572.9    543.6    296.6
##   GSM18928 GSM18929 GSM18930 GSM18931 GSM18932 GSM18933 GSM18934 GSM18935
## 1   2901.3   1857.2   3347.2   2625.7   2532.0   3173.0   4019.9   2662.7
## 2     81.6     16.3     83.4     93.8     18.6     14.6    112.1     69.1
## 3    222.5    195.3    234.1    342.1    169.6    130.3    255.9    394.9
## 4   2457.2   1585.6   1709.5   1855.0   1618.3   1183.5   1854.4   2054.8
## 5    321.2    130.0    161.2    173.2    158.8    192.3    195.5    138.9
## 6    424.6    307.9    445.3    551.3    526.5    285.1    443.9    508.1
##   GSM18936 GSM18937 GSM18938 GSM18939 GSM18940 GSM18941 GSM18942 GSM18943
## 1   2727.0   2119.1   1769.4   2550.2   2158.4   2694.2   2497.1   4398.5
## 2     25.9    136.2     18.9     17.2    153.0     29.0    190.7    108.8
## 3    276.6    411.8     73.2    317.9    367.9    239.3    409.2    341.1
## 4   1289.8   1412.5   2304.5   2885.6   2003.1   2591.7   3007.3   1614.5
## 5    102.3    321.6    296.3    332.8    216.3    374.5    308.8    104.7
## 6    479.7    449.6    416.0    882.8    407.5    957.9    650.1    350.0
##   GSM18944 GSM18945 GSM18946 GSM18947 GSM18948 GSM18949 GSM18950 GSM18951
## 1   4808.6   2141.3   2010.9   1494.8   1319.3   1238.5   1807.9   1240.9
## 2     30.8     18.8     97.4      9.7     30.1    140.3     86.9    112.5
## 3    330.1    188.8    205.0    229.1     48.3    203.2    167.3     90.8
## 4   1882.9   1484.9   1442.7   2284.5   2051.4   1455.4   1347.1   2623.5
## 5    160.7    141.4    147.7    130.0    102.4     57.1     83.0    201.2
## 6    396.0    310.6    267.7    629.5    457.9    442.8    414.7    377.8
##   GSM18952 GSM18953 GSM18954 GSM18955 GSM18956 GSM18957 GSM18958 GSM18959
## 1    925.5   1084.8    691.6   1982.6   1786.9   2476.4   1822.6    717.9
## 2    127.0    106.3     24.5     13.1     46.3     20.2     62.0     52.8
## 3    132.8    301.7    286.0    238.1    160.3    181.1    145.4    108.9
## 4   1778.8   3045.8   1471.3   5618.5   6562.1   1715.6   1250.7   1376.9
## 5     98.4    134.9    114.0    116.0    132.5    142.5     75.8    103.9
## 6    560.3    665.4    653.0    870.8    597.4    326.6    549.3    244.5
##   GSM18960 GSM18961 GSM18962 GSM18963 GSM18964 GSM18965 GSM18966 GSM18967
## 1    817.6   2152.8   2450.7   1632.0   1734.4    874.3    822.7   2202.5
## 2     15.5     85.9     88.8     48.6     53.8     64.0     46.0     79.2
## 3    114.8     70.6    154.3    214.0     93.3    222.7    261.8    165.1
## 4   1298.4   7545.0   6722.3   8545.9   9865.6   1147.4    935.1   1734.1
## 5     38.3     83.9     76.2     90.8     46.8     26.6     35.3    102.8
## 6    557.0    410.7    269.9    391.6    437.6    276.1    276.3    235.2
##   GSM18968 GSM18969 GSM18970 GSM18971 GSM18972 GSM18973 GSM18974 GSM18975
## 1   2383.5    469.5    889.9    417.3    374.9    984.3   1183.0    476.8
## 2     13.1     73.7    161.5     60.8     41.6     48.1     27.1     12.7
## 3    268.3    105.2    202.6     64.0     45.0     21.5     47.1    170.6
## 4   1415.8    924.1   2068.1    787.0    647.0    491.2    645.3   1480.7
## 5     86.9     67.5    137.7     49.6     40.5     27.0     39.8     85.9
## 6    383.6    212.9    472.2    255.6    232.3    119.6     84.4    331.5
##   GSM18976 GSM18977 GSM18978 GSM18979 GSM18980 GSM18981 GSM18982 GSM18983
## 1    408.9   1083.1    903.0    940.6   2153.8   1071.2   1029.1   1249.0
## 2     60.3     18.1     46.6     38.8     29.7     95.4     67.1     54.3
## 3    140.5    137.1    189.1     91.9    104.7    322.5    326.8    317.5
## 4   1224.4   1324.6   1678.1   1345.9   1196.6   2295.8   2019.3   2608.5
## 5     15.7      4.6     34.7     77.6     19.8    560.1    584.6    430.7
## 6    426.7    218.7    345.6    210.9    243.3    330.8    389.2    617.4
##   GSM18984 GSM18985 GSM18986 GSM18987 GSM18988 GSM18989 GSM18990 GSM18991
## 1   1433.6    997.3    656.2    902.5    809.9   1302.8   1062.4   1061.3
## 2    159.8    120.4    151.9     75.7     24.9     24.2    101.6      6.8
## 3    233.9    262.5    238.8    278.5    465.6    404.1    309.1     78.7
## 4   2923.3   2187.5   1549.6   2372.4   1900.4   2860.4   2122.6    990.1
## 5    378.0    584.2    120.1    537.2    596.3    602.4    299.2     80.9
## 6    692.6    514.2    309.2    541.9    302.0    432.7    440.5    446.5
##   GSM18992 GSM18993 GSM18994 GSM18995 GSM18996 GSM18997 GSM18998 GSM18999
## 1    760.0   2448.7   1484.5    878.1    813.7    353.1    439.1   1161.8
## 2     93.8     36.0     13.1     72.6     33.7     37.2      8.6    139.2
## 3    122.6    177.8    186.9    165.8    286.5    158.7     87.2    334.4
## 4   1055.1   2063.2   1373.0   1912.1   1939.8   1344.2   1126.5   2506.8
## 5    132.9     65.3    128.7    145.3    280.8     64.6    117.2    192.1
## 6    263.4    373.2    480.1    680.1    535.7    206.7    197.5    496.9
##   GSM19000 GSM19001 GSM19002 GSM19003 GSM19004 GSM19005 GSM19006 GSM19007
## 1    761.0   1233.3   1468.4   1201.7   1655.4   1481.1   2163.5   1310.4
## 2     35.2     51.2     34.2     53.1    193.7     46.7    137.5     37.6
## 3     84.9    248.2    123.6    491.9    317.2    915.7    788.7    902.9
## 4   2278.1   2459.3   2234.4   3817.7   3342.6   4585.7   5395.7   3728.7
## 5    401.6    180.7    188.4    384.2    541.8    640.4    882.2    441.7
## 6    334.8    376.1    542.4    984.5   1168.3    862.2   1015.1    901.8
##   GSM19008 GSM19009 GSM19010 GSM19011 GSM19012 GSM19013 GSM19014 GSM19015
## 1   1205.3   1564.5   1461.3   2925.4   3369.9   1657.9   1322.6    732.2
## 2     37.9     83.3    126.7    163.9     50.3     52.4    143.8     77.6
## 3    838.6    470.8    456.1   1297.6    708.1    616.0    658.0    107.7
## 4   4631.8   3678.2   3594.2   8048.9   7283.1   4009.4   3086.5   1574.5
## 5    426.9    226.2    471.3   1114.2   1115.9    343.7    285.7    159.4
## 6   1124.5    649.7   1710.2   2061.3   1637.6    901.2   1008.4    330.3
##   GSM19016 GSM19017 GSM19018 GSM19019 GSM19020 GSM19021 GSM19022
## 1    841.7   1557.3   1506.3   1791.0   1697.2    923.0   1078.4
## 2     42.6     95.2     16.9     17.9     82.9     23.2     13.6
## 3    438.1    228.1    201.1    383.3    327.9    309.4    420.6
## 4   1720.2   2405.3   1707.4   1695.3   1317.5   1430.1   1734.7
## 5    266.0    105.8    190.9     40.2    154.2     61.9    111.3
## 6    487.9    459.3    487.8    535.3    458.6    363.0    307.2</code></pre>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="current-working-directory.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="scan.html" class="navigation navigation-next " aria-label="Next page""><i class="fa fa-angle-right"></i></a>

<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": {}
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["intro-to-r.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
