<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Introduction to R</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="This is the complementary manual for QFAB’s Introduction to R workshop">
  <meta name="generator" content="bookdown 0.1.4 and GitBook 2.6.7">

  <meta property="og:title" content="Introduction to R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is the complementary manual for QFAB’s Introduction to R workshop" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Introduction to R" />
  
  <meta name="twitter:description" content="This is the complementary manual for QFAB’s Introduction to R workshop" />
  

<meta name="author" content="QFAB">


<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="current-working-directory.html">
<link rel="next" href="scan.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #0000ff; } /* Keyword */
code > span.ch { color: #008080; } /* Char */
code > span.st { color: #008080; } /* String */
code > span.co { color: #008000; } /* Comment */
code > span.ot { color: #ff4000; } /* Other */
code > span.al { color: #ff0000; } /* Alert */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #008000; font-weight: bold; } /* Warning */
code > span.cn { } /* Constant */
code > span.sc { color: #008080; } /* SpecialChar */
code > span.vs { color: #008080; } /* VerbatimString */
code > span.ss { color: #008080; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #0000ff; } /* ControlFlow */
code > span.op { } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ff4000; } /* Preprocessor */
code > span.do { color: #008000; } /* Documentation */
code > span.an { color: #008000; } /* Annotation */
code > span.cv { color: #008000; } /* CommentVar */
code > span.at { } /* Attribute */
code > span.in { color: #008000; } /* Information */
</style>

<link rel="stylesheet" href="css\style.css" type="text/css" />
</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Abstract</a></li>
<li class="chapter" data-level="1" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>1</b> Getting started</a><ul>
<li class="chapter" data-level="1.1" data-path="how-to-read-this-book.html"><a href="how-to-read-this-book.html"><i class="fa fa-check"></i><b>1.1</b> How to read this book</a></li>
<li class="chapter" data-level="1.2" data-path="rstudio-ide.html"><a href="rstudio-ide.html"><i class="fa fa-check"></i><b>1.2</b> RStudio IDE</a></li>
<li class="chapter" data-level="1.3" data-path="alternatives-to-rstudio.html"><a href="alternatives-to-rstudio.html"><i class="fa fa-check"></i><b>1.3</b> Alternatives to RStudio</a></li>
<li class="chapter" data-level="1.4" data-path="the-r-workspace.html"><a href="the-r-workspace.html"><i class="fa fa-check"></i><b>1.4</b> The R workspace</a></li>
<li class="chapter" data-level="1.5" data-path="literate-programming.html"><a href="literate-programming.html"><i class="fa fa-check"></i><b>1.5</b> Literate programming</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-variables-objects-in-r.html"><a href="data-variables-objects-in-r.html"><i class="fa fa-check"></i><b>2</b> Data variables (objects) in R</a><ul>
<li class="chapter" data-level="2.1" data-path="declaring-variables.html"><a href="declaring-variables.html"><i class="fa fa-check"></i><b>2.1</b> Declaring variables</a></li>
<li class="chapter" data-level="2.2" data-path="datatypes.html"><a href="datatypes.html"><i class="fa fa-check"></i><b>2.2</b> Datatypes</a><ul>
<li class="chapter" data-level="2.2.1" data-path="datatypes.html"><a href="datatypes.html#numeric"><i class="fa fa-check"></i><b>2.2.1</b> Numeric</a></li>
<li class="chapter" data-level="2.2.2" data-path="datatypes.html"><a href="datatypes.html#integer"><i class="fa fa-check"></i><b>2.2.2</b> Integer</a></li>
<li class="chapter" data-level="2.2.3" data-path="datatypes.html"><a href="datatypes.html#logical"><i class="fa fa-check"></i><b>2.2.3</b> Logical</a></li>
<li class="chapter" data-level="2.2.4" data-path="datatypes.html"><a href="datatypes.html#character"><i class="fa fa-check"></i><b>2.2.4</b> Character</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>2.3</b> Functions</a><ul>
<li class="chapter" data-level="2.3.1" data-path="functions.html"><a href="functions.html#nesting-functions"><i class="fa fa-check"></i><b>2.3.1</b> Nesting functions</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>2.4</b> Data structures</a><ul>
<li class="chapter" data-level="2.4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>2.4.1</b> Vectors</a></li>
<li class="chapter" data-level="2.4.2" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>2.4.2</b> Lists</a></li>
<li class="chapter" data-level="2.4.3" data-path="data-structures.html"><a href="data-structures.html#changing-vectors-and-lists"><i class="fa fa-check"></i><b>2.4.3</b> Changing Vectors and Lists</a></li>
<li class="chapter" data-level="2.4.4" data-path="data-structures.html"><a href="data-structures.html#factors"><i class="fa fa-check"></i><b>2.4.4</b> Factors</a></li>
<li class="chapter" data-level="2.4.5" data-path="data-structures.html"><a href="data-structures.html#matrices"><i class="fa fa-check"></i><b>2.4.5</b> Matrices</a></li>
<li class="chapter" data-level="2.4.6" data-path="data-structures.html"><a href="data-structures.html#data-frame"><i class="fa fa-check"></i><b>2.4.6</b> Data frame</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="finding-describing-and-removing-objects.html"><a href="finding-describing-and-removing-objects.html"><i class="fa fa-check"></i><b>2.5</b> Finding, describing and removing objects</a></li>
<li class="chapter" data-level="2.6" data-path="best-practise-and-variable-naming-conventions.html"><a href="best-practise-and-variable-naming-conventions.html"><i class="fa fa-check"></i><b>2.6</b> Best practise and variable naming conventions</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="going-further-with-r.html"><a href="going-further-with-r.html"><i class="fa fa-check"></i><b>3</b> Going further with R</a><ul>
<li class="chapter" data-level="3.1" data-path="documentation.html"><a href="documentation.html"><i class="fa fa-check"></i><b>3.1</b> Documentation</a></li>
<li class="chapter" data-level="3.2" data-path="extending-r-with-packages.html"><a href="extending-r-with-packages.html"><i class="fa fa-check"></i><b>3.2</b> Extending R with packages</a></li>
<li class="chapter" data-level="3.3" data-path="vignettes.html"><a href="vignettes.html"><i class="fa fa-check"></i><b>3.3</b> Vignettes</a><ul>
<li class="chapter" data-level="3.3.1" data-path="vignettes.html"><a href="vignettes.html#citations"><i class="fa fa-check"></i><b>3.3.1</b> Citations</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="sessioninfo.html"><a href="sessioninfo.html"><i class="fa fa-check"></i><b>3.4</b> SessionInfo</a></li>
<li class="chapter" data-level="3.5" data-path="other-ways-of-running-r.html"><a href="other-ways-of-running-r.html"><i class="fa fa-check"></i><b>3.5</b> Other ways of running R</a><ul>
<li class="chapter" data-level="3.5.1" data-path="other-ways-of-running-r.html"><a href="other-ways-of-running-r.html#preparing-an-r-script"><i class="fa fa-check"></i><b>3.5.1</b> Preparing an R script</a></li>
<li class="chapter" data-level="3.5.2" data-path="other-ways-of-running-r.html"><a href="other-ways-of-running-r.html#running-an-r-script"><i class="fa fa-check"></i><b>3.5.2</b> Running an R script</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="importing-and-exporting-data-from-r.html"><a href="importing-and-exporting-data-from-r.html"><i class="fa fa-check"></i><b>4</b> Importing and exporting data from R</a><ul>
<li class="chapter" data-level="4.1" data-path="saving-session-data.html"><a href="saving-session-data.html"><i class="fa fa-check"></i><b>4.1</b> Saving session data</a></li>
<li class="chapter" data-level="4.2" data-path="current-working-directory.html"><a href="current-working-directory.html"><i class="fa fa-check"></i><b>4.2</b> Current working directory</a><ul>
<li class="chapter" data-level="4.2.1" data-path="current-working-directory.html"><a href="current-working-directory.html#file-paths"><i class="fa fa-check"></i><b>4.2.1</b> File paths</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="reading-and-writing-tabular-data.html"><a href="reading-and-writing-tabular-data.html"><i class="fa fa-check"></i><b>4.3</b> Reading and writing tabular data</a></li>
<li class="chapter" data-level="4.4" data-path="scan.html"><a href="scan.html"><i class="fa fa-check"></i><b>4.4</b> Scan</a></li>
<li class="chapter" data-level="4.5" data-path="reading-from-a-web-connection.html"><a href="reading-from-a-web-connection.html"><i class="fa fa-check"></i><b>4.5</b> Reading from a web connection</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="transforming-and-summarising-data-in-r.html"><a href="transforming-and-summarising-data-in-r.html"><i class="fa fa-check"></i><b>5</b> Transforming and summarising data in R</a><ul>
<li class="chapter" data-level="5.1" data-path="user-defined-functions.html"><a href="user-defined-functions.html"><i class="fa fa-check"></i><b>5.1</b> User defined functions</a></li>
<li class="chapter" data-level="5.2" data-path="for-loops-while-loops-and-vectorisation.html"><a href="for-loops-while-loops-and-vectorisation.html"><i class="fa fa-check"></i><b>5.2</b> For loops, While loops and Vectorisation</a></li>
<li class="chapter" data-level="5.3" data-path="apply-lapply.html"><a href="apply-lapply.html"><i class="fa fa-check"></i><b>5.3</b> Apply, Lapply</a></li>
<li class="chapter" data-level="5.4" data-path="basic-data-analysis-methods.html"><a href="basic-data-analysis-methods.html"><i class="fa fa-check"></i><b>5.4</b> Basic data analysis methods</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-presentation-and-reporting.html"><a href="data-presentation-and-reporting.html"><i class="fa fa-check"></i><b>6</b> Data presentation and reporting</a><ul>
<li class="chapter" data-level="6.1" data-path="native-r-methods.html"><a href="native-r-methods.html"><i class="fa fa-check"></i><b>6.1</b> Native R methods</a><ul>
<li class="chapter" data-level="6.1.1" data-path="native-r-methods.html"><a href="native-r-methods.html#boxplots"><i class="fa fa-check"></i><b>6.1.1</b> Boxplots</a></li>
<li class="chapter" data-level="6.1.2" data-path="native-r-methods.html"><a href="native-r-methods.html#histograms"><i class="fa fa-check"></i><b>6.1.2</b> Histograms</a></li>
<li class="chapter" data-level="6.1.3" data-path="native-r-methods.html"><a href="native-r-methods.html#bar-charts"><i class="fa fa-check"></i><b>6.1.3</b> Bar charts</a></li>
<li class="chapter" data-level="6.1.4" data-path="native-r-methods.html"><a href="native-r-methods.html#scatterplots"><i class="fa fa-check"></i><b>6.1.4</b> Scatterplots</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="advanced-plotting-using-ggplot2.html"><a href="advanced-plotting-using-ggplot2.html"><i class="fa fa-check"></i><b>6.2</b> Advanced plotting using ggplot2</a><ul>
<li class="chapter" data-level="6.2.1" data-path="advanced-plotting-using-ggplot2.html"><a href="advanced-plotting-using-ggplot2.html#boxplots-1"><i class="fa fa-check"></i><b>6.2.1</b> Boxplots</a></li>
<li class="chapter" data-level="6.2.2" data-path="advanced-plotting-using-ggplot2.html"><a href="advanced-plotting-using-ggplot2.html#histograms-1"><i class="fa fa-check"></i><b>6.2.2</b> Histograms</a></li>
<li class="chapter" data-level="6.2.3" data-path="advanced-plotting-using-ggplot2.html"><a href="advanced-plotting-using-ggplot2.html#bar-charts-1"><i class="fa fa-check"></i><b>6.2.3</b> Bar charts</a></li>
<li class="chapter" data-level="6.2.4" data-path="advanced-plotting-using-ggplot2.html"><a href="advanced-plotting-using-ggplot2.html#scatterplots-1"><i class="fa fa-check"></i><b>6.2.4</b> Scatterplots</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html"><i class="fa fa-check"></i>Appendix 1</a><ul>
<li class="chapter" data-level="" data-path="summary-of-object-types.html"><a href="summary-of-object-types.html"><i class="fa fa-check"></i>Summary of Object Types</a></li>
<li class="chapter" data-level="" data-path="more-information-on-r-and-bioconductor-for-genomics.html"><a href="more-information-on-r-and-bioconductor-for-genomics.html"><i class="fa fa-check"></i>More information on R and Bioconductor for Genomics</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-2-solutions-to-exercises.html"><a href="appendix-2-solutions-to-exercises.html"><i class="fa fa-check"></i>Appendix 2 - Solutions to exercises</a><ul>
<li class="chapter" data-level="" data-path="using-vectors.html"><a href="using-vectors.html"><i class="fa fa-check"></i>2.4.1 Using vectors</a></li>
<li class="chapter" data-level="" data-path="using-lists.html"><a href="using-lists.html"><i class="fa fa-check"></i>2.4.3 Using lists</a></li>
<li class="chapter" data-level="" data-path="adding-factor-values.html"><a href="adding-factor-values.html"><i class="fa fa-check"></i>2.4.4 Adding factor values</a></li>
<li class="chapter" data-level="" data-path="creating-a-data-frame.html"><a href="creating-a-data-frame.html"><i class="fa fa-check"></i>2.4.6 Creating a data frame</a></li>
<li class="chapter" data-level="" data-path="reading-in-the-geneatlas-data.html"><a href="reading-in-the-geneatlas-data.html"><i class="fa fa-check"></i>4.2 Reading in the GeneAtlas data</a></li>
<li class="chapter" data-level="" data-path="creating-and-using-functions.html"><a href="creating-and-using-functions.html"><i class="fa fa-check"></i>5.1 Creating and using functions</a></li>
<li class="chapter" data-level="" data-path="loops-and-vectorisation.html"><a href="loops-and-vectorisation.html"><i class="fa fa-check"></i>5.2 Loops and vectorisation</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-3-the-sessioninfo-for-this-version-of.html"><a href="appendix-3-the-sessioninfo-for-this-version-of.html"><i class="fa fa-check"></i>Appendix 3 - The SessionInfo for this version of</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="reading-and-writing-tabular-data" class="section level2">
<h2><span class="header-section-number">4.3</span> Reading and writing tabular data</h2>
<p>Data is most commonly shared in tabular formats (Excel presents data in a tabular format). These are most-often text files that contain columns of data separated with comma, tab or other field-delimiters. In R there are some simple methods for quickly reading in massive amounts of data.</p>
<p>For the next few steps in this course we are going to explore a public dataset of human tissue gene expression data that was originally produced by Novartis. The dataset is known as the Novartis GeneAtlas data (now called BioGPS) and can be found in the GEO database under accession GSE1133 (<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE1133" class="uri">http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE1133</a>). We will look at some Affymetrix gene expression data from the study. These data have been prepared from microarray hybridisations using the Affymetrix Human Genome U133A Array. The data is in the file <code>data/Intro_to_R/GSE1133-GPL96_series_matrix.txt</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read data from the training server</span>
data.path &lt;-<span class="st"> &quot;data/Intro_to_R&quot;</span>

<span class="co"># Of course, you will need a different path if the data is on your local P.C.</span>
<span class="co"># e.g. data.path &lt;- &quot;C:/Intro_to_R/Data&quot;</span>
<span class="co"># Create a variable pointing to the file location.</span>

gene.atlas.file &lt;-<span class="st"> </span><span class="kw">file.path</span>(data.path, <span class="st">&quot;GSE1133-GPL96_series_matrix.txt&quot;</span>)

<span class="co"># Then look at the start and finish of that file</span>
<span class="kw">head</span>(<span class="kw">readLines</span>(gene.atlas.file))
<span class="kw">tail</span>(<span class="kw">readLines</span>(gene.atlas.file))</code></pre></div>
<p><strong>What is happening here?</strong></p>
<table>
<colgroup>
<col width="15%" />
<col width="84%" />
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>readLines()</code></td>
<td>reads each line from a target file into a vector. See also <code>read.table</code>, <code>read.csv</code> functions.</td>
</tr>
<tr class="even">
<td><code>file.path()</code></td>
<td><em>DOES NOT</em> read in the contents of a file. It just point to the location of a files on your system. You need to use <code>readLines</code>, <code>read.table</code> or similar to access the contents of that file.</td>
</tr>
</tbody>
</table>
<p>Using the <code>head</code> and <code>tail</code> commands helps us understand the nature of the file. It seems that the <code>!</code> prefix indicates metadata rather than data.</p>
<div class="rmdcaution">
<p>
Beware of printing large variables with RStudio Server, you can freeze your session. Judiciously use methods to limit the output:
</p>
<ul>
<li>
read into a variable first
</li>
<li>
limit the output
</li>
<li>
use <code>head()</code> / <code>tail()</code> functions
</li>
</ul>
</div>
<p>We have started to explore the contents of the expression data file using the code above, but at the moment have not imported it into a variable in R. Before we get to that stage, we will first load the annotation data associated with this expression dataset.</p>
<div class="rmdexercise">
<p>
<strong>Reading in tabular data</strong>
</p>
<ol style="list-style-type: decimal">
<li>
<p>
If you have not already done so, use the previous code chunk to set up your data.path and take an initial look at the expression data in the variable <code>gene.atlas.file</code>.
</p>
</li>
<li>
<p>
Follow the annotated code below to use <code>read.table()</code> to import the annotation data, or metadata, for this experiment
</p>
</li>
</ol>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># As before, define the location of the file</span>
annotation.file &lt;-<span class="st"> </span><span class="kw">file.path</span>(data.path,<span class="st">&quot;GSK_RNA.sdrf&quot;</span>)

<span class="co"># Look at the first six lines of the data</span>
<span class="co"># sep = &quot;\t&quot; means that the data is tab-separated</span>
<span class="kw">head</span>(<span class="kw">read.table</span>(annotation.file, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>))

<span class="co"># The last command included the header as the first line of data.</span>
<span class="co"># Specify (header = TRUE) that the first line contains header information</span>
<span class="kw">head</span>(<span class="kw">read.table</span>(annotation.file, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header=</span><span class="ot">TRUE</span>))

<span class="co"># Use read.table to import the data into an variable</span>
expressionMetadata &lt;-<span class="st"> </span><span class="kw">read.table</span>(annotation.file, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header=</span><span class="ot">TRUE</span>)

<span class="co"># Find out some information about the structure and contents of that new variable</span>
<span class="kw">tail</span>(expressionMetadata)
<span class="kw">dim</span>(expressionMetadata)
<span class="kw">colnames</span>(expressionMetadata)
<span class="kw">str</span>(expressionMetadata$Material.Type)

<span class="co"># By default, read.table imports text fields as factors. That&#39;s probably</span>
<span class="co"># not what we want, so disable it with stringsAsFactors=FALSE</span>
expressionMetadata &lt;-<span class="st"> </span><span class="kw">read.table</span>(annotation.file, 
                                 <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header=</span><span class="ot">TRUE</span>, 
                                 <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
<span class="kw">str</span>(expressionMetadata$Material.Type)</code></pre></div>
<p>Note that the lines starting with the default comment symbol (#) have been omitted from the import.</p>
<div class="rmdexercise">
<p>
<strong>Read the expression data into a variable called <code>gene.atlas</code></strong>
</p>
<p>
In the last couple of sections we have used <code>readLines()</code> to look at the content of a file and we have used <code>read.table()</code> to read some experimental metadata into a data frame. Your challenge is now to import the “GSE1133-GPL96_series_matrix.txt” data into a data frame variable called <code>gene.atlas</code>.
</p>
<p>
<em>Hints</em>:
</p>
<ul>
<li>
Use the <code>read.table</code> documentation to discover the nuances of the function.
</li>
<li>
This import is not as simple as the annotation file. You will need to make some inferences about the structure of the file (<em>Hint</em>: use <code>head()</code> and <code>tail()</code> and look at some of the options for <code>read.table()</code> in the Help.)
</li>
<li>
Make sure that the proper column names are included in the data frame.
</li>
</ul>
<p>
The output should look something like the following where only the first 6 rows and 6 columns are shown.
</p>
</div>
<pre><code>##           GSM18865 GSM18866 GSM18867 GSM18868 GSM18869 GSM18870
## 1007_s_at   2082.0   1577.2    231.8    328.4     83.6     73.0
## 1053_at      598.4    256.2     57.0    102.1     39.1     43.6
## 117_at       194.2    135.9    512.3    713.4    108.6     75.7
## 121_at      1910.5   1516.2    529.8    990.2    383.4    396.4
## 1255_g_at    138.4     53.8     30.7     75.7     16.9     22.8
## 1294_at      263.4    287.2    347.8    345.6    300.8    377.7</code></pre>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="current-working-directory.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="scan.html" class="navigation navigation-next " aria-label="Next page""><i class="fa fa-angle-right"></i></a>

<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"facebook": false,
"twitter": false,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": {}
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["intro-to-r.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
