<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Introduction to R</title>
  <meta name="description" content="This is the complementary manual for QFAB’s Introduction to R workshop">
  <meta name="generator" content="bookdown 0.4 and GitBook 2.6.7">

  <meta property="og:title" content="Introduction to R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is the complementary manual for QFAB’s Introduction to R workshop" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Introduction to R" />
  
  <meta name="twitter:description" content="This is the complementary manual for QFAB’s Introduction to R workshop" />
  

<meta name="author" content="QFAB">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="current-working-directory.html">
<link rel="next" href="scan.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #0000ff; } /* Keyword */
code > span.ch { color: #008080; } /* Char */
code > span.st { color: #008080; } /* String */
code > span.co { color: #008000; } /* Comment */
code > span.ot { color: #ff4000; } /* Other */
code > span.al { color: #ff0000; } /* Alert */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #008000; font-weight: bold; } /* Warning */
code > span.cn { } /* Constant */
code > span.sc { color: #008080; } /* SpecialChar */
code > span.vs { color: #008080; } /* VerbatimString */
code > span.ss { color: #008080; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #0000ff; } /* ControlFlow */
code > span.op { } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ff4000; } /* Preprocessor */
code > span.do { color: #008000; } /* Documentation */
code > span.an { color: #008000; } /* Annotation */
code > span.cv { color: #008000; } /* CommentVar */
code > span.at { } /* Attribute */
code > span.in { color: #008000; } /* Information */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Abstract</a></li>
<li class="chapter" data-level="1" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>1</b> Getting started</a><ul>
<li class="chapter" data-level="1.1" data-path="how-to-read-this-book.html"><a href="how-to-read-this-book.html"><i class="fa fa-check"></i><b>1.1</b> How to read this book</a></li>
<li class="chapter" data-level="1.2" data-path="rstudio-ide.html"><a href="rstudio-ide.html"><i class="fa fa-check"></i><b>1.2</b> RStudio IDE</a></li>
<li class="chapter" data-level="1.3" data-path="alternatives-to-rstudio.html"><a href="alternatives-to-rstudio.html"><i class="fa fa-check"></i><b>1.3</b> Alternatives to RStudio</a></li>
<li class="chapter" data-level="1.4" data-path="the-r-workspace.html"><a href="the-r-workspace.html"><i class="fa fa-check"></i><b>1.4</b> The R workspace</a></li>
<li class="chapter" data-level="1.5" data-path="literate-programming.html"><a href="literate-programming.html"><i class="fa fa-check"></i><b>1.5</b> Literate programming</a></li>
<li class="chapter" data-level="1.6" data-path="documentation.html"><a href="documentation.html"><i class="fa fa-check"></i><b>1.6</b> Documentation</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-variables-objects-in-r.html"><a href="data-variables-objects-in-r.html"><i class="fa fa-check"></i><b>2</b> Data variables (objects) in R</a><ul>
<li class="chapter" data-level="2.1" data-path="declaring-variables.html"><a href="declaring-variables.html"><i class="fa fa-check"></i><b>2.1</b> Declaring variables</a></li>
<li class="chapter" data-level="2.2" data-path="data-types.html"><a href="data-types.html"><i class="fa fa-check"></i><b>2.2</b> Data types</a><ul>
<li class="chapter" data-level="2.2.1" data-path="data-types.html"><a href="data-types.html#numeric-and-integer"><i class="fa fa-check"></i><b>2.2.1</b> Numeric and Integer</a></li>
<li class="chapter" data-level="2.2.2" data-path="data-types.html"><a href="data-types.html#character"><i class="fa fa-check"></i><b>2.2.2</b> Character</a></li>
<li class="chapter" data-level="2.2.3" data-path="data-types.html"><a href="data-types.html#logical"><i class="fa fa-check"></i><b>2.2.3</b> Logical</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="determining-data-type.html"><a href="determining-data-type.html"><i class="fa fa-check"></i><b>2.3</b> Determining data type</a><ul>
<li class="chapter" data-level="2.3.1" data-path="determining-data-type.html"><a href="determining-data-type.html#typecasting"><i class="fa fa-check"></i><b>2.3.1</b> Typecasting</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>2.4</b> Functions</a><ul>
<li class="chapter" data-level="2.4.1" data-path="functions.html"><a href="functions.html#nesting-functions"><i class="fa fa-check"></i><b>2.4.1</b> Nesting functions</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>2.5</b> Data structures</a><ul>
<li class="chapter" data-level="2.5.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>2.5.1</b> Vectors</a></li>
<li class="chapter" data-level="2.5.2" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>2.5.2</b> Lists</a></li>
<li class="chapter" data-level="2.5.3" data-path="data-structures.html"><a href="data-structures.html#changing-vectors-and-lists"><i class="fa fa-check"></i><b>2.5.3</b> Changing Vectors and Lists</a></li>
<li class="chapter" data-level="2.5.4" data-path="data-structures.html"><a href="data-structures.html#factors"><i class="fa fa-check"></i><b>2.5.4</b> Factors</a></li>
<li class="chapter" data-level="2.5.5" data-path="data-structures.html"><a href="data-structures.html#matrices"><i class="fa fa-check"></i><b>2.5.5</b> Matrices</a></li>
<li class="chapter" data-level="2.5.6" data-path="data-structures.html"><a href="data-structures.html#data-frame"><i class="fa fa-check"></i><b>2.5.6</b> Data frame</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="finding-describing-and-removing-objects.html"><a href="finding-describing-and-removing-objects.html"><i class="fa fa-check"></i><b>2.6</b> Finding, describing and removing objects</a></li>
<li class="chapter" data-level="2.7" data-path="best-practise-and-variable-naming-conventions.html"><a href="best-practise-and-variable-naming-conventions.html"><i class="fa fa-check"></i><b>2.7</b> Best practise and variable naming conventions</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="simple-data-analysis.html"><a href="simple-data-analysis.html"><i class="fa fa-check"></i><b>3</b> Simple data analysis</a><ul>
<li class="chapter" data-level="3.1" data-path="useful-operations.html"><a href="useful-operations.html"><i class="fa fa-check"></i><b>3.1</b> Useful operations</a></li>
<li class="chapter" data-level="3.2" data-path="vectorisation.html"><a href="vectorisation.html"><i class="fa fa-check"></i><b>3.2</b> Vectorisation</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="file-and-data-inputoutput-io.html"><a href="file-and-data-inputoutput-io.html"><i class="fa fa-check"></i><b>4</b> File and data input/output (IO)</a><ul>
<li class="chapter" data-level="4.1" data-path="saving-session-data.html"><a href="saving-session-data.html"><i class="fa fa-check"></i><b>4.1</b> Saving session data</a></li>
<li class="chapter" data-level="4.2" data-path="current-working-directory.html"><a href="current-working-directory.html"><i class="fa fa-check"></i><b>4.2</b> Current working directory</a><ul>
<li class="chapter" data-level="4.2.1" data-path="current-working-directory.html"><a href="current-working-directory.html#file-paths"><i class="fa fa-check"></i><b>4.2.1</b> File paths</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="reading-and-writing-tabular-data.html"><a href="reading-and-writing-tabular-data.html"><i class="fa fa-check"></i><b>4.3</b> Reading and writing tabular data</a></li>
<li class="chapter" data-level="4.4" data-path="scan.html"><a href="scan.html"><i class="fa fa-check"></i><b>4.4</b> Scan</a></li>
<li class="chapter" data-level="4.5" data-path="reading-from-a-web-connection.html"><a href="reading-from-a-web-connection.html"><i class="fa fa-check"></i><b>4.5</b> Reading from a web connection</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="packages-and-functions.html"><a href="packages-and-functions.html"><i class="fa fa-check"></i><b>5</b> Packages and functions</a><ul>
<li class="chapter" data-level="5.1" data-path="extending-r-with-packages.html"><a href="extending-r-with-packages.html"><i class="fa fa-check"></i><b>5.1</b> Extending R with packages</a></li>
<li class="chapter" data-level="5.2" data-path="user-defined-functions.html"><a href="user-defined-functions.html"><i class="fa fa-check"></i><b>5.2</b> User defined functions</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="more-complex-data-analysis.html"><a href="more-complex-data-analysis.html"><i class="fa fa-check"></i><b>6</b> More complex data analysis</a><ul>
<li class="chapter" data-level="6.1" data-path="apply.html"><a href="apply.html"><i class="fa fa-check"></i><b>6.1</b> apply()</a></li>
<li class="chapter" data-level="6.2" data-path="for-and-while-loops.html"><a href="for-and-while-loops.html"><i class="fa fa-check"></i><b>6.2</b> For and while loops</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="visualising-data.html"><a href="visualising-data.html"><i class="fa fa-check"></i><b>7</b> Visualising data</a><ul>
<li class="chapter" data-level="7.1" data-path="boxplots.html"><a href="boxplots.html"><i class="fa fa-check"></i><b>7.1</b> Boxplots</a></li>
<li class="chapter" data-level="7.2" data-path="histograms.html"><a href="histograms.html"><i class="fa fa-check"></i><b>7.2</b> Histograms</a></li>
<li class="chapter" data-level="7.3" data-path="scatterplots.html"><a href="scatterplots.html"><i class="fa fa-check"></i><b>7.3</b> Scatterplots</a></li>
<li class="chapter" data-level="7.4" data-path="bar-charts.html"><a href="bar-charts.html"><i class="fa fa-check"></i><b>7.4</b> Bar charts</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html"><i class="fa fa-check"></i>Appendix 1</a><ul>
<li class="chapter" data-level="" data-path="summary-of-object-types.html"><a href="summary-of-object-types.html"><i class="fa fa-check"></i>Summary of Object Types</a></li>
<li class="chapter" data-level="" data-path="more-information-on-r-and-bioconductor-for-genomics.html"><a href="more-information-on-r-and-bioconductor-for-genomics.html"><i class="fa fa-check"></i>More information on R and Bioconductor for Genomics</a></li>
<li class="chapter" data-level="7.5" data-path="going-further-with-r.html"><a href="going-further-with-r.html"><i class="fa fa-check"></i><b>7.5</b> Going further with R</a><ul>
<li class="chapter" data-level="7.5.1" data-path="going-further-with-r.html"><a href="going-further-with-r.html#vignettes"><i class="fa fa-check"></i><b>7.5.1</b> Vignettes</a></li>
<li class="chapter" data-level="7.5.2" data-path="going-further-with-r.html"><a href="going-further-with-r.html#citations"><i class="fa fa-check"></i><b>7.5.2</b> Citations</a></li>
<li class="chapter" data-level="7.5.3" data-path="going-further-with-r.html"><a href="going-further-with-r.html#sessioninfo"><i class="fa fa-check"></i><b>7.5.3</b> SessionInfo</a></li>
<li class="chapter" data-level="7.5.4" data-path="going-further-with-r.html"><a href="going-further-with-r.html#other-ways-of-running-r"><i class="fa fa-check"></i><b>7.5.4</b> Other ways of running R</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="reading-and-writing-tabular-data" class="section level2">
<h2><span class="header-section-number">4.3</span> Reading and writing tabular data</h2>
<p>Data is most commonly shared in tabular formats (Excel presents data in a tabular format). These are most-often text files that contain columns of data separated with comma, tab or other field-delimiters. In R there are some simple methods for quickly reading in massive amounts of data.</p>
<p>For the next few steps in this course we are going to explore a public dataset of human tissue gene expression data that was originally produced by Novartis. The dataset is known as the Novartis GeneAtlas data (now called BioGPS) and can be found in the GEO database under accession GSE1133 (<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE1133" class="uri">http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE1133</a>). We will look at some Affymetrix gene expression data from the study. These data have been prepared from microarray hybridisations using the Affymetrix Human Genome U133A Array. The data is in the file <code>data/Intro_to_R/GSE1133-GPL96_series_matrix.txt</code>.</p>
<div class="figure">
<img src="images/microrarray-data.png" alt="Sample data" />
<p class="caption">Sample data</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read data from the training server</span>
data.path &lt;-<span class="st"> &quot;data/Intro_to_R&quot;</span>

<span class="co"># Of course, you will need a different path if the data is on your local P.C.</span>
<span class="co"># e.g. data.path &lt;- &quot;C:/Intro_to_R/Data&quot;</span>
<span class="co"># Create a variable pointing to the file location.</span>

gene.atlas.file &lt;-<span class="st"> </span><span class="kw">file.path</span>(data.path, <span class="st">&quot;GSE1133-GPL96_series_matrix.txt&quot;</span>)

<span class="co"># Then look at the start and finish of that file</span>
<span class="kw">head</span>(<span class="kw">readLines</span>(gene.atlas.file))
<span class="kw">tail</span>(<span class="kw">readLines</span>(gene.atlas.file))</code></pre></div>
<p><strong>What is happening here?</strong></p>
<table>
<colgroup>
<col width="15%" />
<col width="84%" />
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>readLines()</code></td>
<td>reads each line from a target file into a vector. See also <code>read.table</code>, <code>read.csv</code> functions.</td>
</tr>
<tr class="even">
<td><code>file.path()</code></td>
<td><em>DOES NOT</em> read in the contents of a file. It just point to the location of a files on your system. You need to use <code>readLines</code>, <code>read.table</code> or similar to access the contents of that file.</td>
</tr>
</tbody>
</table>
<p>Using the <code>head</code> and <code>tail</code> commands helps us understand the nature of the file. It seems that the <code>!</code> prefix indicates metadata rather than data.</p>
<div class="rmdcaution">
<p>
Beware of printing large variables with RStudio Server, you can freeze your session. Judiciously use methods to limit the output:
</p>
<ul>
<li>
read into a variable first
</li>
<li>
limit the output
</li>
<li>
use <code>head()</code> / <code>tail()</code> functions
</li>
</ul>
</div>
<p>We have started to explore the contents of the expression data file using the code above, but at the moment <strong>have not imported</strong> it into a variable in R. The following exercise will step you through using the <code>read.table()</code> function in R to read a text file and import the data as a data.frame object.</p>
<p>When reading a table, R expects the same number of columns for every row, it will by default use the first line to determine the number of columns that are in your dataset. You can specify the separater to use depending on how your data columns are separated e.g. tab-delimited, comma-delimited or by something else altogether.</p>
<div class="rmdexercise">
<p>
<strong>Reading in tabular data</strong>
</p>
<ol style="list-style-type: decimal">
<li>
<p>
If you have not already done so, use the previous code chunk to set up your data.path and take an initial look at the expression data in the variable <code>gene.atlas.file</code>.
</p>
</li>
<li>
<p>
Open up the help documentation for <code>read.table()</code> function while you perform the following worked example. Read up on the following parameters in detail:
</p>
</li>
</ol>
<ul>
<li>
<code>sep</code>
</li>
<li>
<code>quote</code>
</li>
<li>
<code>comment</code>
</li>
<li>
<code>header</code>
</li>
<li>
<code>row.names</code>
</li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li>
Follow the code below to use <code>read.table()</code> to import the dataset into R. There should be 22,283 rows (observations) and 158 columns (variables).
</li>
</ol>
</div>
<p>Pay close attention to the dimension of the dataset after each read as each parameter will affect the read.</p>

<p><strong>Attempt 1:: read.table using default settings</strong>:</p>
<p>This will fail because by default the <code>sep</code> parameter separates on white-space and the first X lines in the file are all comments, each line do not have the same number of columns as indicated by the error message.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene.atlas &lt;-<span class="st"> </span><span class="kw">read.table</span>(gene.atlas.file)</code></pre></div>
<p><code># Error in scan(file = file, what = what, sep = sep, quote = quote, dec = dec, : line 48 did not have 2 elements</code></p>

<p><strong>Attempt 2:: skip comment lines</strong></p>
<p>We specify the comment character but check the dimension of the data after reading, is this correct? Also examine the first 5 rows and columns</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene.atlas &lt;-<span class="st"> </span><span class="kw">read.table</span>(gene.atlas.file, <span class="dt">comment.char =</span> <span class="st">&quot;!&quot;</span>)
<span class="kw">dim</span>(gene.atlas)</code></pre></div>
<pre><code>## [1] 22284   159</code></pre>
<p>We have 1 too many columns and rows in our <code>gene.atlas</code> data frame and taking a physical look a the object will tell us why. It is because we haven’t assigned the row and column names to the matrix</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene.atlas[<span class="dv">1</span>:<span class="dv">5</span>,<span class="dv">1</span>:<span class="dv">5</span>]</code></pre></div>
<pre><code>##          V1       V2       V3       V4       V5
## 1    ID_REF GSM18865 GSM18866 GSM18867 GSM18868
## 2 1007_s_at     2082   1577.2    231.8    328.4
## 3   1053_at    598.4    256.2       57    102.1
## 4    117_at    194.2    135.9    512.3    713.4
## 5    121_at   1910.5   1516.2    529.8    990.2</code></pre>

<p><strong>Attempt 3:: assign header and row names</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene.atlas &lt;-<span class="st"> </span><span class="kw">read.table</span>(gene.atlas.file, <span class="dt">comment.char=</span><span class="st">&quot;!&quot;</span>, 
                         <span class="dt">header=</span><span class="ot">TRUE</span>, 
                         <span class="dt">row.names=</span><span class="dv">1</span>)
<span class="kw">dim</span>(gene.atlas)</code></pre></div>
<pre><code>## [1] 22283   158</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene.atlas[<span class="dv">1</span>:<span class="dv">6</span>,<span class="dv">1</span>:<span class="dv">6</span>]</code></pre></div>
<pre><code>##           GSM18865 GSM18866 GSM18867 GSM18868 GSM18869 GSM18870
## 1007_s_at   2082.0   1577.2    231.8    328.4     83.6     73.0
## 1053_at      598.4    256.2     57.0    102.1     39.1     43.6
## 117_at       194.2    135.9    512.3    713.4    108.6     75.7
## 121_at      1910.5   1516.2    529.8    990.2    383.4    396.4
## 1255_g_at    138.4     53.8     30.7     75.7     16.9     22.8
## 1294_at      263.4    287.2    347.8    345.6    300.8    377.7</code></pre>
<p>Much better, this looks right!</p>
<div class="rmdexercise">
<p>
<strong>Optional exercise</strong>
</p>
<p>
If you are ahead, try reading in the associated metadata file for this dataset. The file is called “GSK_RNA.sdrf”. It is a normal text file and since it is less than 1MB in size, you can physically open up the file to have a look first before using <code>read.table()</code>. This will tell you what parameters you need to make use of.
</p>
</div>
<p>Note that the lines starting with the default comment symbol (#) have been omitted from the import.</p>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="current-working-directory.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="scan.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": {}
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["intro-to-r.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
